util/vendor.py

```diff
@@ -340,46 +358,82 @@ def main(argv):
         raise SystemExit(sys.exc_info()[1])
     desc['_base_dir'] = vendor_file_base_dir
 
+    first_import = False
+
     # Load lock file contents (if possible)
-    desc_file_path_str = str(desc_file_path)
-    lock_file_path = Path(
-        desc_file_path_str[:desc_file_path_str.find('.vendor.hjson')] +
-        '.lock.hjson')
+    desc_file_stem = desc_file_path.name.rsplit('.', 2)[0]
+    lock_file_path = desc_file_path.with_name(desc_file_stem + '.lock.hjson')
     try:
-        with open(lock_file_path, 'r') as f:
+        with open(str(lock_file_path), 'r') as f:
             lock = hjson.loads(f.read(), use_decimal=True)
     except FileNotFoundError:
-        log.warning(
-            "Unable to read lock file %s. Assuming this is the first import.",
-            lock_file_path)
         lock = None
+        if args.update:
+            log.fatal("Cannot update without existing lock file %s.",
+                      str(lock_file_path))
+            log.fatal(
+                "Retry without the '--update' flag if this is an initial import."
+            )
+            raise SystemExit(1)
+        else:
+            log.warning(
+                "Unable to read lock file %s. Assuming this is the first import.",
+                str(lock_file_path))
+            first_import = True
 
     if args.refresh_patches:
         refresh_patches(desc)
 
     clone_dir = Path(tempfile.mkdtemp())
     try:
+        if first_import or args.update:
+            url_to_clone = desc['upstream']['url']
+            rev_to_clone = desc['upstream']['rev']
+        else:
+            url_to_clone = lock['upstream']['url']
+            rev_to_clone = lock['upstream']['rev']
+
         # clone upstream repository
-        upstream_new_rev = clone_git_repo(desc['upstream']['url'], clone_dir,
-                                          desc['upstream']['rev'])
+        upstream_new_rev = clone_git_repo(url_to_clone,
+                                          clone_dir,
+                                          rev=rev_to_clone)
+
+        # We can skip this check if we're the first import
+        if first_import:
+            pass
+        elif args.update:
+            if upstream_new_rev == lock['upstream']['rev']:
+                log.fatal("Upstream Revision has not changed.")
+                log.fatal(
+                    "This could be because the rev in %s was not updated",
+                    str(desc_file_path))
+                raise SystemExit(1)
+        else:
+            if upstream_new_rev != rev_to_clone:
+                log.fatal(
+                    "Revision Mismatch. Unable to re-clone locked version of repository."
+                )
+                log.fatal("Locked Version: %s", rev_to_clone)
+                log.fatal("Re-cloned Version: %s", upstream_new_rev)
+                raise SystemExit(1)
```

Comment:
Can you put this check into clone_git_repo?, optionally as an assert? It's a violation of the postcondition of `clone_git_repo`.

Reply:
This cannot go into `clone_git_repo`: if you're using `--update`, the `rev` argument to `clone_git_repo` will likely be something like `master`, but the return value is always a sha. I'll try to work out how to make this check slightly less invasive.